lighthouse-core/audits/service-worker.js

```diff
@@ -73,23 +73,33 @@ class ServiceWorker extends Audit {
    * @param {Array<LH.Crdp.ServiceWorker.ServiceWorkerVersion>} matchingSWVersions
    * @param {Array<LH.Crdp.ServiceWorker.ServiceWorkerRegistration>} registrations
    * @param {URL} pageUrl
-   * @return {string|undefined}
+   * @return {{scopeUrl: string; scriptUrl: string} | undefined}
    */
-  static getControllingScopeUrl(matchingSWVersions, registrations, pageUrl) {
+  static getControllingServiceWorker(matchingSWVersions, registrations, pageUrl) {
     // Find the normalized scope URLs of possibly-controlling SWs.
-    const matchingScopeUrls = matchingSWVersions
-      .map(v => registrations.find(r => r.registrationId === v.registrationId))
-      .filter(/** @return {r is LH.Crdp.ServiceWorker.ServiceWorkerRegistration} */ r => !!r)
-      .map(r => new URL(r.scopeURL).href);
+    /** @type {Array<{ scopeUrl: string; scriptUrl: string;}>} */
+    const scriptByScopeUrlList = [];
+
+    // Populate serviceWorkerUrls map with the scriptURLs and scopeUrls of matchingSWVersions and registrations
+    for (const version of matchingSWVersions) {
+      const matchedRegistration = registrations
+        .find(r => r.registrationId === version.registrationId);
+
+      if (matchedRegistration) {
+        const scopeUrl = new URL(matchedRegistration.scopeURL).href;
+        const scriptUrl = new URL(version.scriptURL).href;
+        scriptByScopeUrlList.push({scopeUrl, scriptUrl});
+      }
+    }
 
     // Find most-specific applicable scope, the one controlling the page.
     // See https://w3c.github.io/ServiceWorker/v1/#scope-match-algorithm
-    const pageControllingScope = matchingScopeUrls
-      .filter(scopeUrl => pageUrl.href.startsWith(scopeUrl))
-      .sort((scopeA, scopeB) => scopeA.length - scopeB.length)
+    const pageControllingUrls = scriptByScopeUrlList
+      .filter(urlPair => pageUrl.href.startsWith(urlPair.scopeUrl.toString()))
```

Comment:
Nit: I believe `urlPair.scopeUrl` is already of type `string` so the `.toString()` isn't needed.